temporalio/contrib/openai_agents/invoke_model_activity.py

```diff
@@ -0,0 +1,187 @@
+    strict_json_schema: bool
+
+    def is_plain_text(self) -> bool:
+        """Whether the output type is plain text (versus a JSON object)."""
+        return self.output_type_name is None or self.output_type_name == "str"
+
+    def is_strict_json_schema(self) -> bool:
+        """Whether the JSON schema is in strict mode."""
+        return self.strict_json_schema
+
+    def json_schema(self) -> dict[str, Any]:
+        """The JSON schema of the output type."""
+        if self.is_plain_text():
+            raise UserError("Output type is plain text, so no JSON schema is available")
+        if self.output_schema is None:
+            raise UserError("Output schema is not defined")
+        return self.output_schema
+
+    def validate_json(self, json_str: str) -> Any:
+        """Validate the JSON string against the schema."""
+        raise NotImplementedError()
+
+    def name(self) -> str:
+        """Get the name of the output type."""
+        if self.output_type_name is None:
+            raise ValueError("output_type_name is None")
+        return self.output_type_name
+
+
+class ModelTracingInput(enum.IntEnum):
+    """Conversion friendly representation of ModelTracing.
+
+    Needed as ModelTracing is enum.Enum instead of IntEnum
+    """
+
+    DISABLED = 0
+    ENABLED = 1
+    ENABLED_WITHOUT_DATA = 2
+
+
+class ActivityModelInput(TypedDict, total=False):
+    """Input for the invoke_model_activity activity."""
+
+    model_name: Optional[str]
+    system_instructions: Optional[str]
+    input: Required[Union[str, list[TResponseInputItem]]]  # type: ignore
+    model_settings: Required[ModelSettings]
+    tools: list[ToolInput]
+    output_schema: Optional[AgentOutputSchemaInput]
+    handoffs: list[HandoffInput]
+    tracing: Required[ModelTracingInput]
+    previous_response_id: Optional[str]
+    prompt: Optional[Any]
+
+
+class ModelActivity:
+    """Class wrapper for model invocation activities to allow model customization."""
+
+    def __init__(self, model_provider: Union[ModelProvider, None] = None):
+        """Initialize the activity with a model provider."""
+        self._model_provider = model_provider or MultiProvider()
+
+    @activity.defn
+    @_auto_heartbeater
+    async def invoke_model_activity(self, input: ActivityModelInput) -> ModelResponse:
+        """Activity that invokes a model with the given input."""
+        with workflow.unsafe.sandbox_unrestricted():
+            model = self._model_provider.get_model(input.get("model_name"))
+
+        async def empty_on_invoke_tool(ctx: RunContextWrapper[Any], input: str) -> str:
+            return ""
+
+        async def empty_on_invoke_handoff(
+            ctx: RunContextWrapper[Any], input: str
+        ) -> Any:
+            return None
+
+        # workaround for https://github.com/pydantic/pydantic/issues/9541
+        # ValidatorIterator returned
+        input_json = json.dumps(input["input"], default=lambda o: str(o))
```

Comment:
Nit: should be equivalent
```suggestion
        input_json = json.dumps(input["input"], default=str)
```